### PR TITLE
feat: support the timeSpecification X.509v3 extension

### PR DESCRIPTION
**Pull request by JonathanWilbur**
This PR adds support for encoding, decoding, and displaying the `timeSpecification` X.509v3 extension defined in ITU-T Recommendation X.509 for usage in attribute certificates.

##### Checklist
- [x] documentation is added or updated
- [x] tests are added or updated

##### Syntax

The syntax of this extension is:

```asn
timeSpecification EXTENSION ::= {
  SYNTAX         TimeSpecification
  IDENTIFIED BY  id-ce-timeSpecification }

TimeSpecification ::= SEQUENCE {
  time           CHOICE {
    absolute       SEQUENCE {
      startTime [0]  GeneralizedTime OPTIONAL,
      endTime   [1]  GeneralizedTime OPTIONAL,
      ... },
    periodic      SET SIZE (1..MAX) OF Period},
  notThisTime   BOOLEAN DEFAULT FALSE,
  timeZone      TimeZone OPTIONAL,
  ... }

Period ::= SEQUENCE {
  timesOfDay  [0]  SET SIZE (1..MAX) OF DayTimeBand OPTIONAL,
  days        [1]  CHOICE {
    intDay           SET OF INTEGER,
    bitDay           BIT STRING {
      sunday    (0),
      monday    (1),
      tuesday   (2),
      wednesday (3),
      thursday  (4),
      friday    (5),
      saturday  (6)},
    dayOf            XDayOf,
    ...} OPTIONAL,
  weeks       [2]  CHOICE {
    allWeeks         NULL,
    intWeek          SET OF INTEGER,
    bitWeek          BIT STRING {
      week1     (0),
      week2     (1),
      week3     (2),
      week4     (3),
      week5     (4)},
    ... } OPTIONAL,
  months      [3]  CHOICE {
    allMonths        NULL,
    intMonth         SET OF INTEGER,
    bitMonth         BIT STRING {
      january   (0),
      february  (1),
      march     (2),
      april     (3),
      may       (4),
      june      (5),
      july      (6),
      august    (7),
      september (8),
      october   (9),
      november  (10),
      december  (11)},
    ...} OPTIONAL,
  years       [4]  SET OF INTEGER(1000..MAX) OPTIONAL,
  ... }

XDayOf ::= CHOICE {
  first   [1]  NamedDay,
  second  [2]  NamedDay,
  third   [3]  NamedDay,
  fourth  [4]  NamedDay,
  fifth   [5]  NamedDay }

NamedDay ::= CHOICE {
  intNamedDays ENUMERATED {
    sunday      (1),
    monday      (2),
    tuesday     (3),
    wednesday   (4),
    thursday    (5),
    friday      (6),
    saturday    (7)},
  bitNamedDays BIT STRING {
    sunday      (0),
    monday      (1),
    tuesday     (2),
    wednesday   (3),
    thursday    (4),
    friday      (5),
    saturday    (6)} }

DayTimeBand ::= SEQUENCE {
  startDayTime  [0]  DayTime DEFAULT {hour 0},
  endDayTime    [1]  DayTime DEFAULT {hour 23, minute 59, second 59},
  ... }

DayTime ::= SEQUENCE {
  hour    [0]  INTEGER(0..23),
  minute  [1]  INTEGER(0..59) DEFAULT 0,
  second  [2]  INTEGER(0..59) DEFAULT 0,
  ... }

TimeZone ::= INTEGER(-12..12)

TimeAssertion ::= CHOICE {
  now             NULL,
  at              GeneralizedTime,
  between         SEQUENCE {
    startTime  [0]  GeneralizedTime,
    endTime    [1]  GeneralizedTime OPTIONAL,
    entirely        BOOLEAN DEFAULT FALSE,
    ...},
  ... }
```

### CODE REVIEW

**Comment by JonathanWilbur** [comment]
For the sake of previewing what the output looks like, the output of the supplied test certificates is:

```
# ./apps/openssl x509 -noout -text -in ./test/certs/ext-timeSpecification-absolute.pem 
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 203569230 (0xc22384e)
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: CN=Hi mom
        Validity
            Not Before: Dec 20 13:07:21 2022 GMT
            Not After : Dec 20 13:07:21 2022 GMT
        Subject: CN=Hi mom
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:b6:78:cb:9b:5b:6c:d6:10:b8:7c:d3:6d:dd:49:
                    d0:0f:dc:bb:dd:b0:d7:82:2a:13:c9:66:12:23:77:
                    1a:fc:a3:5f:b9:3c:9d:8c:54:c4:60:09:aa:9e:e1:
                    ab:04:e5:1f:81:b9:87:dc:f4:50:d0:09:a9:aa:58:
                    e0:59:36:f7:77:72:bc:1f:85:05:bc:35:90:4c:f4:
                    92:db:52:0c:8e:6f:28:a9:dd:7d:9c:4e:46:c5:68:
                    ab:1a:ed:28:0d:1c:e1:58:b1:de:d5:16:7b:66:ba:
                    21:08:f8:3f:d4:e7:2c:cb:8e:d0:66:e2:b6:95:a1:
                    41:d2:b1:26:ad:15:81:98:46:1b:0e:5d:ff:c3:17:
                    f9:5d:fa:88:aa:93:b2:9c:90:74:da:d8:57:ad:30:
                    9c:4c:ce:d1:cf:51:6a:08:5b:7f:65:56:41:e6:12:
                    98:d9:2f:82:4d:d1:3d:38:85:4a:96:be:16:1c:c7:
                    ee:bd:46:1e:3a:3d:3a:1a:4c:0b:14:91:cd:bf:6b:
                    54:fa:da:2f:23:79:98:47:15:2e:63:85:2e:a6:44:
                    b7:30:2f:8e:b5:ec:bb:5c:a9:b5:3f:e2:0d:8d:65:
                    a8:7a:09:ba:88:20:2d:dd:5b:ac:a5:5c:fe:ea:95:
                    36:c6:09:77:9e:b0:15:31:01:c1:e1:f4:5e:40:f1:
                    f4:a5
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Time Specification: 
                Timezone: UTC-05:00
                Time:
                    Absolute: Any time between Dec 20 13:07:21 2022 GMT and Dec 20 13:07:21 2022 GMT

    Signature Algorithm: sha1WithRSAEncryption
    Signature Value:

# ./apps/openssl x509 -noout -text -in ./test/certs/ext-timeSpecification-periodic.pem 
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 203569230 (0xc22384e)
        Signature Algorithm: sha1WithRSAEncryption
        Issuer: CN=Hi mom
        Validity
            Not Before: Dec 21 14:49:42 2022 GMT
            Not After : Dec 21 14:49:42 2022 GMT
        Subject: CN=Hi mom
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:b6:78:cb:9b:5b:6c:d6:10:b8:7c:d3:6d:dd:49:
                    d0:0f:dc:bb:dd:b0:d7:82:2a:13:c9:66:12:23:77:
                    1a:fc:a3:5f:b9:3c:9d:8c:54:c4:60:09:aa:9e:e1:
                    ab:04:e5:1f:81:b9:87:dc:f4:50:d0:09:a9:aa:58:
                    e0:59:36:f7:77:72:bc:1f:85:05:bc:35:90:4c:f4:
                    92:db:52:0c:8e:6f:28:a9:dd:7d:9c:4e:46:c5:68:
                    ab:1a:ed:28:0d:1c:e1:58:b1:de:d5:16:7b:66:ba:
                    21:08:f8:3f:d4:e7:2c:cb:8e:d0:66:e2:b6:95:a1:
                    41:d2:b1:26:ad:15:81:98:46:1b:0e:5d:ff:c3:17:
                    f9:5d:fa:88:aa:93:b2:9c:90:74:da:d8:57:ad:30:
                    9c:4c:ce:d1:cf:51:6a:08:5b:7f:65:56:41:e6:12:
                    98:d9:2f:82:4d:d1:3d:38:85:4a:96:be:16:1c:c7:
                    ee:bd:46:1e:3a:3d:3a:1a:4c:0b:14:91:cd:bf:6b:
                    54:fa:da:2f:23:79:98:47:15:2e:63:85:2e:a6:44:
                    b7:30:2f:8e:b5:ec:bb:5c:a9:b5:3f:e2:0d:8d:65:
                    a8:7a:09:ba:88:20:2d:dd:5b:ac:a5:5c:fe:ea:95:
                    36:c6:09:77:9e:b0:15:31:01:c1:e1:f4:5e:40:f1:
                    f4:a5
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Time Specification: 
                Timezone: UTC-05:00
                NOT this time:
                    Periodic:
                        Period:
                            Daytime bands:
                                05:43:21 - 12:34:56

                            Days of the week: SUN, MON
                            Weeks of the month: 3, 4
                            Months: MAY, JUN
                            Years: 2022, 2023
                        Period:
                            Days of the month: 3, 4
                            Months: JUL, AUG
                            Years: 2023, 2024

    Signature Algorithm: sha1WithRSAEncryption
    Signature Value:
```

**Comment by JonathanWilbur** [comment]
This PR is ready for review now.

The coding style failures appear to be unresolvable or make no sense. In the former case, they fail because they both want single statements to not appear in a code block, but they also want `else` conditions to appear in a code block if the non-`else` condition appears in a code block. In the latter case, they are failing on `x509v3.h.in` on a part where I am generating stack macros using the exact same syntax as is used elsewhere. The error itself makes no sense to me.

The other failure appears to be some sort of outage unrelated to my changes.

**Review by t8m — APPROVED** [review]

**Comment by JonathanWilbur** [comment]
@nhorman would you care to review?

**Comment by JonathanWilbur** [comment]
@Sashan would you care to review this one?

**Comment by JonathanWilbur** [comment]
Linting is probably going to fail as described above. My last commit removed the extra newline after `Daytime bands` in the output copied above.

**Review by t8m — APPROVED** [review]

**Comment by JonathanWilbur** [comment]
@vdukhovni would you care to review this one?

**Comment by JonathanWilbur** [comment]
@mattcaswell would you care to review this?

**Comment by JonathanWilbur** [comment]
I fixed the merge conflict. This is ready for review.

**Review by t8m — APPROVED** [review]

**Comment by JonathanWilbur** [comment]
@nhorman would you care to review this one?

File: crypto/x509/v3_timespec.c [hunk 1]
```diff
@@ -0,0 +1,754 @@
...
+/*
+ * It might seem like you could just print the bits of the string numerically,
+ * but the fifth bit has the special meaning of "the final week" imputed to it
+ * by the text of ITU-T Recommendation X.520.
+ */
+static int print_bit_week(BIO *out, ASN1_BIT_STRING *bs)
+{
+    int i = 0;
+
+    if (ASN1_BIT_STRING_get_bit(bs, OSSL_TIME_SPEC_BIT_WEEKS_1)) {
```
Comment (nhorman) [inline crypto/x509/v3_timespec.c:297]: Might it be better to write this as a for loop to de-duplicate all the get_bit and BIO_puts calls, and just skip bit 5 with a continue statement?

Comment (JonathanWilbur) [reply crypto/x509/v3_timespec.c]: Done. That was a much more elegant approach! Thank you!

**Comment by JonathanWilbur** [comment]
This is ready for review again.

**Review by nhorman — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**